types/parse/parse-tests.ts

```diff
@@ -1325,145 +1329,192 @@ function testObject() {
         // $ExpectType void
         objUntyped.revert();
 
         // $ExpectType void
-        objUntyped.revert('whatever', 'more whatever');
+        objUntyped.revert("whatever", "more whatever");
 
         // $ExpectType void
         objTyped.revert();
 
         // $ExpectType void
-        objTyped.revert('thingOne', 'thingTwo');
+        objTyped.revert("thingOne", "thingTwo");
 
         // $ExpectError
-        objTyped.revert('other');
+        objTyped.revert("other");
+    }
+    interface ObjectAttributes {
+        example: boolean;
+        someString: string;
+    }
+    interface OptionalObjectAttributes {
+        example?: boolean;
+        another?: string;
     }
 
     async function testSave(
         objUntyped: Parse.Object,
-        objTyped: Parse.Object<{ example: boolean; someString: string }>,
+        objTyped: Parse.Object<ObjectAttributes>,
+        objTypedOptional: Parse.Object<OptionalObjectAttributes>,
     ) {
         // $ExpectType Object<Attributes>
         await objUntyped.save({ whatever: 100 });
 
         // $ExpectType Object<Attributes>
-        await objUntyped.save('whatever', 100);
+        await objUntyped.save("whatever", 100);
 
-        // $ExpectType Object<{ example: boolean; someString: string; }>
+        // $ExpectType Object<ObjectAttributes>
         await objTyped.save({ example: true });
 
-        // $ExpectType Object<{ example: boolean; someString: string; }>
-        await objTyped.save({ example: true, someString: 'hello' });
+        // $ExpectType Object<ObjectAttributes>
+        await objTyped.save({ example: true, someString: "hello" });
 
         // $ExpectError
-        await objTyped.save({ example: 'hello', someString: true });
+        await objTyped.save({ example: "hello", someString: true });
 
-        // $ExpectType Object<{ example: boolean; someString: string; }>
-        await objTyped.save('example', true);
+        // $ExpectType Object<ObjectAttributes>
+        await objTyped.save("example", true);
 
         // $ExpectError
-        await objTyped.save({ example: 'hello' });
+        await objTyped.save({ example: "hello" });
 
         // $ExpectError
         await objTyped.save({ wrongProp: 5 });
 
         // $ExpectError
-        await objTyped.save('example', 10);
+        await objTyped.save("example", 10);
+
+        // $ExpectError
+        await objTyped.save("wrongProp", true);
+
+        /**
+         * Parse doesn't throw when a value is `undefined`,
+         * but saves won't be taken into account.
+         */
+
+        // $ExpectError
+        await objTyped.save({ example: undefined });
+
+        // $ExpectError
+        await objTyped.save("example", undefined);
+
+        // $ExpectError
+        await objTypedOptional.save({ example: undefined });
 
         // $ExpectError
-        await objTyped.save('wrongProp', true);
+        await objTypedOptional.save("example", undefined);
```

Comment:
Ah...I misinterpreted what you were saying in your other comment...so actually in these cases, since the attributes type explicitly types the properties as `undefined`, we would have allowed setting these. I do see how it could be beneficial to prevent ever setting a value to be `undefined` in order to force/guide users more towards `.unset`...but I do think that may be a bit heavy-handed.

More and more, I think the core issue for your case anyway is some difference between the behavior of `class Thing extends Parse.Object<SomeAttributes>` and `new Parse.Object<SomeAttributes>('Thing')`. It seems like the extended class is losing some TS context somehow...

Reply:
I would vote that we support these test cases:
```ts
        // $ExpectError
        await objTyped.save({ example: undefined });

        // $ExpectError
        await objTyped.save("example", undefined);

        // $ExpectType Object<OptionalObjectAttributes>
        await objTypedOptional.save({ example: undefined });

        // $ExpectType Object<OptionalObjectAttributes>
        await objTypedOptional.save("example", undefined);
```

Which, by removing the `Required<>` from your `save()` and `set()` types, I believe should all pass and also still support Intellisense for all kinds of attribute types, partial or not.